x-pack/apm-server/sampling/processor_test.go

```diff
@@ -60,18 +60,31 @@ func TestProcessAlreadyTailSampled(t *testing.T) {
 	// subsequent events in the trace will be reported immediately.
 	trace1 := model.Trace{ID: "0102030405060708090a0b0c0d0e0f10"}
 	trace2 := model.Trace{ID: "0102030405060708090a0b0c0d0e0f11"}
-	storage := eventstorage.New(config.DB, eventstorage.JSONCodec{}, time.Minute, 0)
+	storage := eventstorage.New(config.DB, eventstorage.JSONCodec{})
 	writer := storage.NewReadWriter()
-	assert.NoError(t, writer.WriteTraceSampled(trace1.ID, true))
-	assert.NoError(t, writer.Flush())
+	wOpts := eventstorage.WriterOpts{
+		TTL:                 time.Minute,
+		StorageLimitInBytes: 0,
+	}
+	assert.NoError(t, writer.WriteTraceSampled(trace1.ID, true, wOpts))
+	assert.NoError(t, writer.Flush(wOpts.StorageLimitInBytes))
 	writer.Close()
 
-	storage = eventstorage.New(config.DB, eventstorage.JSONCodec{}, -1, 0) // expire immediately
+	wOpts.TTL = -1
+	storage = eventstorage.New(config.DB, eventstorage.JSONCodec{})
 	writer = storage.NewReadWriter()
-	assert.NoError(t, writer.WriteTraceSampled(trace2.ID, true))
-	assert.NoError(t, writer.Flush())
+	assert.NoError(t, writer.WriteTraceSampled(trace2.ID, true, wOpts))
+	assert.NoError(t, writer.Flush(wOpts.StorageLimitInBytes))
 	writer.Close()
 
+	// Overwrite global storage to allow the transactions created by
+	// sharded read writers to view the explicitly sampled traces so far
+	eventCodec := eventstorage.JSONCodec{}
+	config.Storage = eventstorage.
+		New(config.DB, eventCodec).
+		NewShardedReadWriter()
+	t.Cleanup(func() { config.Storage.Close() })
```

Comment:
IIUC, this is necessary because `config.Storage` is created before the transactions are written, so they are not visible to the storage's transaction due to SSI. Could you add a bit more detail to the comment to help others (and future me :smile:) understand the rationale?

Maybe instead of replacing the storage, we could just flush it? That will cause new transactions to be created.

Reply:
> Could you add a bit more detail to the comment to help others (and future me 😄) understand the rationale?

Thanks, will help my future self to save some confusion too 😄 

> Maybe instead of replacing the storage, we could just flush it?

Yeah, this will work too and will be cleaner—I will update.